Introduce NIP-44 encryption standard (#715)
nostr-protocol/nips: 44.md

NIP-44
======

Encrypted Payloads (Versioned)
------------------------------

`optional` `author:paulmillr` `author:staab`

The NIP introduces a new data format for keypair-based encryption. This NIP is versioned to allow multiple algorithm choices to exist simultaneously.

> An encrypted payload MUST be encoded as a JSON object. Different versions may have different parameters. Every format has a `v` field specifying its version.

Reviewer: ```suggestion
The encrypted payload is encoded as base64. The first byte of the payload represents the version.
```

Currently defined encryption algorithms:

- `0x00` - Reserved
- `0x01` - XChaCha with same key `sha256(ecdh)` per conversation

# Version 1

Params:

1. `nonce`: base64-encoded xchacha nonce
2. `ciphertext`: base64-encoded xchacha ciphertext, created from (key, nonce) against `plaintext`.

> 

Reviewer: Something like this?

- Bob's private key: `4b22aa260e4acb7021e32f38a6cdf4b673c6a277755bfce287e370c924dc936d`

Encrypting the message `hello` from Alice to Bob results in the base-64 encoded tlv payload:

```
AAEBARgeI8gcP/4mnw3mKgtMvD8aGYUnGBlhopoCBd94Ev9i
```

# Other Notes

By default in the [libsecp256k1](https://github.com/bitcoin-core/secp256k1) ECDH implementation, the secret is the SHA256 hash of the shared point (both X and Y coordinates). We are using this exact implementation. In NIP-94, unhashed shared point was used.

This encryption scheme replaces the one described in NIP-04, which is not secure. It used bad cryptographic building blocks and must not be used.

# Code Samples

## Javascript

```javascript
import {base64} from '@scure/base'
import {randomBytes} from '@noble/hashes/utils'
import {secp256k1} from '@noble/curves/secp256k1'
import {sha256} from '@noble/hashes/sha256'
import {xchacha20} from '@noble/ciphers/chacha'

import {utf8Decoder, utf8Encoder} from './utils.ts'

export const getSharedSecret = (privkey: string, pubkey: string): Uint8Array =>
  sha256(secp256k1.getSharedSecret(privkey, '02' + pubkey).subarray(1, 33))

> export function encrypt(privkey: string, pubkey: string, text: string, v = 1) {

Reviewer 2: Should it be `privkey: string, pubkey: string`, or rather `key: Uint8Array`?

Caching an encryption key somewhere is definitely a goal. Using `privkey, pubkey` signature makes it impossible.

Reviewer 2: No, user can't re-derive encryption key for 50 messages. That would consume all the performance. ECDH is slow, like, 500 per sec on M2. They need to derive key once and use the same key for encryption / decryption of all conversation messages, and there can be 1000s.

Reviewer 2: ```js
const msgs = await fetchMessages(100);

// original
let key = getSharedSecret(privA, pubB);
msgs.map(msg => decrypt(key, msg)); // pure decryption routine: 100K/sec

// vs your proposal
msgs.map(msg => decrypt(privA, pubB, msg)); // decryption + getSharedSecret: 200x slower
```

This is just one convo. There can be like 20.

Author: This sounds like an optimization, couldn't you just lru_cache the `getSharedSecret` function?

Reviewer 2: One frame is best-case 16.6ms, worst-case 8.3ms. Decrypting messages in main thread (when you can't use separate one) would just make UI stuck. Even with a separate thread, decrypting 500 conversation messages could take 2-5 seconds on a mobile phone with your proposal. Forcing users to wait is terrible UX.

Author: Got it, yes the base version should just take a key, I'll update that. For stuff like browser extensions, there's no need to expose the key-only version (if the application has a key they can import nostr-tools or whatever), so that interface will look a little different.

NIP-44
======

Encrypted Payloads (Versioned)
------------------------------

`optional` `author:paulmillr` `author:staab`

The NIP introduces a new data format for keypair-based encryption. This NIP is versioned to allow multiple algorithm choices to exist simultaneously.

> The encrypted payload is encoded as base64. The first byte of the payload represents the version. The rest depends on the version.

Reviewer 3: base64 without padding?

Reviewer: By padding you mean the `===` stuff at the end?

Author: I was wondering the same thing. From what I read, base64 padding is to ensure the right number of bytes, extra padding seems unusual.

Reviewer 3: I meant the == stuff. Some base64 impls use it, some don't. Looks like we are using it. Good, that's more normal.

This is very different from padding the plaintext which I talk about elsewhere.

Currently defined encryption algorithms:

- `0x00` - Reserved

> - `0x01` - XChaCha with same key `sha256(ecdh)` per conversation

Reviewer 3: XChaCha20 (add the '20' to the end)

Also what is this "per conversation"?

Reviewer 2: It's common to call it just ChaCha, instead of Salsa20 - which is usually suffixed with 20. So it's cool I think.

Reviewer 3: But there are 8 and 12 round versions. Crypto people may know that 20 is implied, but nostr devs may stumble upon the other ones and not know which one to use. See this lib for example: https://docs.rs/chacha20/latest/chacha20/

# Version 1

Encoding:

- `1` - 1 byte, version
- `nonce` - 24 bytes fetched from a crytographically secure random source
- `ciphertext` - variable number of bytes until the end

> `ciphertext` is created using `xchacha(key, nonce, plaintext)`

Reviewer 3: xchacha20(...)   (again the 20 should be there)

Reviewer 3: Also at this point you should explain what 'key' is, rather than putting that in a note later, because this is where the thing is being defined.

Example:

- Alice's private key: `5c0c523f52a5b6fad39ed2403092df8cebc36318b39383bca6c00808626fab3a`
- Bob's private key: `4b22aa260e4acb7021e32f38a6cdf4b673c6a277755bfce287e370c924dc936d`

Encrypting the message `hello` from Alice to Bob results in the following base-64 encoded payload:

```

> Ac4wiXzHQ20IrriCl8ZNzPWEGre97jKP6THD4ACR

Reviewer 3: This cannot be verified unless you tell us what you used for a random 24-byte nonce.

Also, I think there ought to be a test vector going in the other direction too.

Reviewer: That's true, but I think it's fine just as a soft-example of what it looks like.

Reviewer 3: I recant my statement. The nonce is in the output.

  val nonce = ByteArray(24).apply {
    SecureRandom.getInstanceStrong().nextBytes(this)
  }

  val cipher = streamXChaCha20Xor(
    message = msg.toByteArray(),
    nonce = nonce,
    key = getSharedSecretNIP44(privKey, pubKey)
  )

  return Base64.getEncoder().encodeToString(
    byteArrayOf(Nip44Version.XChaCha20.code) + nonce + cipher
  )
}

fun decryptNIP44(base64: String, privKey: ByteArray, pubKey: ByteArray): String? {
  val byteArray = Base64.getDecoder().decode(base64)

  require(byteArray[0].toInt() == Nip44Version.XChaCha20.code) { "NIP44: unknown encryption version" }

  return streamXChaCha20Xor(
    message = byteArray.copyOfRange(25, byteArray.size),
    nonce = byteArray.copyOfRange(1, 25),
    key = getSharedSecretNIP44(privKey, pubKey)
  )?.decodeToString()
}

// This method is not exposed in AndroidSodium yet, but it will be in the next version.
fun streamXChaCha20Xor(message: ByteArray, nonce: ByteArray, key: ByteArray): ByteArray? {
  return with (SodiumAndroid()) {
    val resultCipher = ByteArray(message.size)

    val isSuccessful = crypto_stream_chacha20_xor_ic(
      resultCipher,
      message,
      message.size.toLong(),

>       nonce.drop(16).toByteArray(), // chacha nonce is just the last 8 bytes.

Reviewer 2: How does this work?

Reviewer 4: Not on libsodium. That code is from them.

Reviewer 4: https://github.com/jedisct1/libsodium/blob/4dc02ce8414e70b8960bea77abf4fe46efb99e93/src/libsodium/crypto_stream/xchacha20/stream_xchacha20.c#L45

Reviewer 2: yeah, it's ok, same as in js then.

NIP-44
======

Encrypted Payloads (Versioned)
------------------------------

`optional` `author:paulmillr` `author:staab`

The NIP introduces a new data format for keypair-based encryption. This NIP is versioned to allow multiple algorithm choices to exist simultaneously.

The encrypted payload is encoded as base64. The first byte of the payload represents the version. The rest depends on the version.

Currently defined encryption algorithms:

- `0x00` - Reserved
- `0x01` - XChaCha with same key `sha256(ecdh)` per conversation

# Version 1

Encoding:

- `1` - 1 byte, version

> - `nonce` - 24 bytes

Reviewer 2: it's important to note somewhere, perhaps here, that nonce should be fetched from CSPRNG (cryptographically secure random source). If we leave it up to users, they will use garbage and leak their messages through nonce reuse.

NIP-44
======

Encrypted Payloads (Versioned)
------------------------------

`optional` `author:paulmillr` `author:staab`

> The NIP introduces a versioned encryption data model, allowing multiple algorithm choices to exist simultaneously.

Reviewer 5: I would suggest introducing the topic a little better, and then providing some clarity about where and how this should be used.

```suggestion
Nostr can use private key to encrypt data for a particular user (pubkey). Then, that user can decrypt the data with their own private key. This NIP introduces a versioned JSON format for symmetrically encrypted data, allowing multiple algorithm choices to exist simultaneously.

The resulting JSON data could be used in the content of an event, or in a tag. This NIP defines only the data model and supported algorithms and not the events themselves.
```

Author: I don't think we need to go quite back to basics, but I've tweaked the wording based on your suggestions. Also, this is generally for asymmetric rather than symmetric encryption.

> The algorithm described in NIP-04 is potentially vulnerable to [padding oracle attacks](https://en.wikipedia.org/wiki/Padding_oracle_attack) and uses keys which are not indistinguishable from random. For more information, see [here](https://github.com/nostr-protocol/nips/pull/715#issuecomment-1675301250-).

Reviewer 5: This link has an accidental `-` at the end.

```suggestion
The algorithm described in NIP-04 is potentially vulnerable to [padding oracle attacks](https://en.wikipedia.org/wiki/Padding_oracle_attack) and uses keys which are not indistinguishable from random. For more information, see [here](https://github.com/nostr-protocol/nips/pull/715#issuecomment-1675301250).
```

NIP-44
======

> Encrypted Direct Message (Versioned)

Reviewer 4: Encryption Payload Definition

------------------------------------

`optional` `author:paulmillr` `author:staab`

> The NIP introduces versioned encryption, allowing multiple algorithm choices to exist simultaneously.

Reviewer 4: The NIP introduces a versioned encryption data model, allowing multiple algorithm choices to exist simultaneously.

> The algorithm described in NIP4 is potentially vulnerable to [padding oracle attacks](https://en.wikipedia.org/wiki/Padding_oracle_attack) and uses keys which are not indistinguishable from random.

Reviewer 4: Delete

Author: I think this is helpful context for now.

An encrypted payload MUST be encoded as a JSON object. Different versions may have different parameters. Every format has a `v` field specifying its version.

Currently defined encryption algorithms:

- `0x00` - Reserved
- `0x01` - XChaCha with same key `sha256(ecdh)` per conversation

# Version 0

> Version 0 is not defined, however implementations depending on this NIP MAY choose to support the payload described in NIP 04 in the same places a NIP 44 payload would otherwise be expected. This is intended to allow a smooth transition while clients and signing software adopt the new standard.

Reviewer 4: I don't like this ambiguity. We should say v0 is the NIP04-styled json. @v0l what you think?

Reviewer 2: nip4 is broken cryptography that will expose your private keys. It should never be used.

  "nonce": "3dBKd83Pg2Q4Tu2A2e8N++c+ZW2IBc2f",
  "v": 1
}
```

**Note**: By default in the [libsecp256k1](https://github.com/bitcoin-core/secp256k1) ECDH implementation, the secret is the SHA256 hash of the shared point (both X and Y coordinates). We are using this exact implementation. In NIP4, unhashed shared point was used.

## Code Samples

### Javascript

```javascript
import {xchacha20} from "@noble/ciphers/chacha"
import {secp256k1} from "@noble/curves/secp256k1"
import {sha256} from "@noble/hashes/sha256"
import {randomBytes} from "@noble/hashes/utils"
import {base64} from "@scure/base"

export const utf8Decoder = new TextDecoder()

export const utf8Encoder = new TextEncoder()

export const getSharedSecret = (privkey: string, pubkey: string): Uint8Array =>
  sha256(secp256k1.getSharedSecret(privkey, "02" + pubkey).subarray(1, 33))

export function encrypt(privkey: string, pubkey: string, text: string, v = 1) {
  if (v !== 1) {
    throw new Error("NIP44: unknown encryption version")
  }

  const key = getSharedSecret(privkey, pubkey)
  const nonce = randomBytes(24)
  const plaintext = utf8Encoder.encode(text)
  const ciphertext = xchacha20(key, nonce, plaintext)

>   return JSON.stringify({
>     ciphertext: base64.encode(ciphertext),
>     nonce: base64.encode(nonce),
>     v,
>   })

Reviewer 6: I'd prefer this so we can include it in at tag:

```suggestion
  return [base64.encode(ciphertext), base64.encode(nonce), v]
```

but if its too much bike shedding at this point than so be it.

Author: I'll resolve this conversation since it's a duplicate of https://github.com/nostr-protocol/nips/pull/716#discussion_r1291530129

NIP-44
======

Encrypted Payloads (Versioned)
------------------------------

`optional` `author:paulmillr` `author:staab`

The NIP introduces a new data format for keypair-based encryption. This NIP is versioned to allow multiple algorithm choices to exist simultaneously.

The encrypted payload is encoded as base64 using the character set defined in RFC4648. The first byte of the payload represents the version. The rest depends on the version.

Currently defined encryption algorithms:

- `0x00` - Reserved

> - `0x01` - XChaCha20 with same key `sha256(parity + ecdh)` per conversation

Reviewer 3: Actually the parity byte is part of the output of ecdh for most libraries, so this would put it in twice.

I would put something like `ecdh(private_key, public_key) where the ecdh() function computes the shared point, and then returns the sha256 hash of the x-only component prefixed by the parity byte ('02').`

Reviewer 2: except that the libraries implement ECDSA and not schnorr. There is no parity byte in schnorr.

Reviewer 3: I only understand a little. I thought the parity byte represented the Y coordinate and there were two options since the curve loops back around on itself. And I know that with Schnorr signatures it is always even (2) for some reason.  And I thought DH just did successive multiplications to come up to the same point  g^a * g^b = g^b * g^a and I don't understand where either signature algorithm (DSA or Schnorr) comes in when determining the DH shared point (and does ECDSA even have a shared point?).

All I want is for the NIP to do the default thing that libraries do, rather than a custom thing that requires lower-level operations.

# Version 1

Encoding:

- `1` - 1 byte, version

> - `nonce` - 24 bytes fetched from a crytographically secure random source

Reviewer 7: ```suggestion
- `nonce` - 24 bytes fetched from a cryptographically secure random source
```

_insert "it's not much but it's honest work" meme_

The NIP introduces a versioned encryption data model, allowing multiple algorithm choices to exist simultaneously.

> The algorithm described in NIP-04 is potentially vulnerable to [padding oracle attacks](https://en.wikipedia.org/wiki/Padding_oracle_attack) and uses keys which are not indistinguishable from random. For more information, see [here](https://github.com/nostr-protocol/nips/pull/715#issuecomment-1675301250).

Reviewer 2: Don't think it's good idea to link to the comment. Can we simplify it?

> The algorithm described in NIP-04 is not secure. It used bad cryptographic building blocks and must not be used.

Author: Your comment was so helpful, but yeah we have it in this PR for reference. I don't think most people realize that NIP 04 compromises your private key.

## Javascript

```javascript
import {base64} from '@scure/base'
import {randomBytes} from '@noble/hashes/utils'
import {secp256k1} from '@noble/curves/secp256k1'
import {sha256} from '@noble/hashes/sha256'
import {xchacha20} from '@noble/ciphers/chacha'

import {utf8Decoder, utf8Encoder} from './utils.ts'

export const getSharedSecret = (privkey: string, pubkey: string): Uint8Array =>

>   sha256(secp256k1.getSharedSecret(privkey, '02' + pubkey).subarray(1, 33))

Reviewer 2: My only concern with regards to the algorithms in the NIP is the fact we're always prepending `02` parity byte. Not sure about potential consequences of that. E.g. bias towards one, or another kind of shared keys.
